crates/provider/src/layers/gas.rs

```diff
@@ -0,0 +1,301 @@
+///
+/// You cannot construct this directly, use [`ProviderBuilder`] with a [`GasEstimatorLayer`].
+///
+/// [`ProviderBuilder`]: crate::ProviderBuilder
+#[derive(Debug, Clone)]
+pub struct GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    /// Gets the gas_price to be used in legacy txs.
+    async fn get_gas_price(&self) -> TransportResult<U256> {
+        self.inner.get_gas_price().await
+    }
+
+    /// Gets the gas_limit to be used in txs.
+    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<U256> {
+        self.inner.estimate_gas(tx, None).await
+    }
+
+    /// Gets the max_fee_per_gas and max_priority_fee_per_gas to be used in EIP-1559 txs.
+    async fn get_eip1559_fees_estimate(&self) -> TransportResult<(U256, U256)> {
+        self.inner.estimate_eip1559_fees(None).await
+    }
+
+    /// Populates the gas_limit, max_fee_per_gas and max_priority_fee_per_gas fields if unset.
+    /// Requires the chain_id to be set in the transaction request to be processed as a EIP-1559 tx.
+    /// If the network does not support EIP-1559, it will process it as a legacy tx.
+    async fn handle_eip1559_tx<'a, 'b>(
+        &'a self,
+        tx: &'b mut N::TransactionRequest,
+    ) -> Result<&'b mut N::TransactionRequest, TransportError> {
+        let gas_estimate = self.get_gas_estimate(tx);
+        let eip1559_fees = self.get_eip1559_fees_estimate();
+
+        let (gas_estimate, eip1559_fees) = futures::join!(gas_estimate, eip1559_fees);
+
+        gas_estimate.map(|gas_estimate| tx.set_gas_limit(gas_estimate))?;
+
+        match eip1559_fees {
+            Ok((max_fee_per_gas, max_priority_fee_per_gas)) => {
+                tx.set_max_fee_per_gas(max_fee_per_gas);
+                tx.set_max_priority_fee_per_gas(max_priority_fee_per_gas);
+                Ok(tx)
+            }
+            Err(err) => {
+                if err.is_transport_error()
+                    && err.to_string() == *"EIP-1559 not activated".to_string()
+                {
+                    // If EIP-1559 is not activated, it will process as a legacy tx.
+                    match self.handle_legacy_tx(tx).await {
+                        Ok(tx) => Ok(tx),
+                        Err(err) => Err(err),
+                    }
+                } else {
+                    Err(err)
+                }
+            }
+        }
+    }
+
+    /// Populates the gas_price and only populates the gas_limit field if unset.
+    /// This method always assumes that the gas_price is unset.
+    async fn handle_legacy_tx<'a, 'b>(
+        &'a self,
+        tx: &'b mut N::TransactionRequest,
+    ) -> Result<&'b mut N::TransactionRequest, TransportError> {
+        let gas_price = self.get_gas_price();
```

Comment:
here's a nice little pattern for future usage:

```
        let gas_limit_fut = if let Some(gas_limit) = tx.gas_limit() {
            futures::future::Either::Left(async move { gas_limit })
        } else {
            futures::future::Either::Right(async { self.estimate_gas(&*tx, None).await })
        };
```

Reply:
with FuturesExt trait

```
let gas_limit_fut = tx.gas_limit() {
            async move { gas_limit }.left_future()
        } else {
            async { self.estimate_gas(&*tx, None).await }.right_future()
        };
```

Reply:
that way you can run the `join!` without all the clunky if block

Reply:
This was a really cool thing I learned! Thanks.